README.md
[![Build Status](https://travis-ci.org/dotandimet/Mojo-UserAgent-Role-Queued.svg?branch=master)](https://travis-ci.org/dotandimet/Mojo-UserAgent-Role-Queued) [![MetaCPAN Release](https://badge.fury.io/pl/Mojo-UserAgent-Role-Queued.svg)](https://metacpan.org/release/Mojo-UserAgent-Role-Queued)
# NAME

Mojo::UserAgent::Role::Queued - A role to process non-blocking requests in a rate-limiting queue.

# SYNOPSIS

       use Mojo::UserAgent;

       my $ua = Mojo::UserAgent->new->with_role('+Queued');
       $ua->max_redirects(3);
       $ua->max_active(5); # process up to 5 requests at a time
       for my $url (@big_list_of_urls) {
       $ua->get($url, sub {
               my ($ua, $tx) = @_;
               if ($tx->success) {
                   say "Page at $url is titled: ",
                     $tx->res->dom->at('title')->text;
               }
              });
      };
      # works with promises, too:
     my @p = map {
       $ua->get_p($_)->then(sub { pop->res->dom->at('title')->text })
         ->catch(sub { say "Error: ", @_ })
     } @big_list_of_urls;
      Mojo::Promise->all(@p)->wait;
    

# DESCRIPTION

Mojo::UserAgent::Role::Queued manages all non-blocking requests made through [Mojo::UserAgent](https://metacpan.org/pod/Mojo::UserAgent) in a queue to limit the number of simultaneous requests.

[Mojo::UserAgent](https://metacpan.org/pod/Mojo::UserAgent) can make multiple concurrent non-blocking HTTP requests using Mojo's event loop, but because there is only a single process handling all of them, you must take care to limit the number of simultaneous requests you make.

Some discussion of this issue is available here
[http://blogs.perl.org/users/stas/2013/01/web-scraping-with-modern-perl-part-1.html](http://blogs.perl.org/users/stas/2013/01/web-scraping-with-modern-perl-part-1.html)
and in Joel Berger's answer here:
[http://stackoverflow.com/questions/15152633/perl-mojo-and-json-for-simultaneous-requests](http://stackoverflow.com/questions/15152633/perl-mojo-and-json-for-simultaneous-requests).

[Mojo::UserAgent::Role::Queued](https://metacpan.org/pod/Mojo::UserAgent::Role::Queued) tries to generalize the practice of managing a large number of requests using a queue, by embedding the queue inside [Mojo::UserAgent](https://metacpan.org/pod/Mojo::UserAgent) itself.

# EVENTS

[Mojo::UserAgent::Role::Queued](https://metacpan.org/pod/Mojo::UserAgent::Role::Queued) adds the following event to those emitted by [Mojo::UserAgent](https://metacpan.org/pod/Mojo::UserAgent):

## queue\_empty

    $ua->on(queue_empty => sub { my ($ua) = @_; .... })

Emitted when the queue has been emptied of all pending jobs. In previous releases, this event was called `stop_queue` (**this is a breaking change**).

# ATTRIBUTES

[Mojo::UserAgent::Role::Queued](https://metacpan.org/pod/Mojo::UserAgent::Role::Queued) has the following attributes:

## max\_active

    $ua->max_active(5);  # execute no more than 5 transactions at a time.
    print "Execute no more than ", $ua->max_active, " concurrent transactions"

Parameter controlling the maximum number of transactions that can be active at the same time.

## 

# LICENSE

Copyright (C) Dotan Dimet.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Dotan Dimet <dotan@corky.net>


Changes
Revision history for Perl extension Mojo-UserAgent-Role-Queued

{{$NEXT}}

1.15 2018-11-15T16:40:28Z
    - correct required Mojolicious version to v7.56 (for -role)

1.14 2018-10-07T15:44:25Z

    - rename queue attribute to job_queue (prevent clash with Mojo::UA->{queue})
    - (try to) fix memory leaks (thanks to Vidar Tyldum)
    - prevent test failures (thanks to CPAN Testers)

1.13 2018-09-27T10:44:41Z
    - make tests that depend on extra modules optional, to reduce CPANTesters fails

1.12 2018-09-26T14:30:24Z
    - fix memory leak (Vidar Tyldum)

1.11 2018-09-12T14:09:24Z
    - change queue from state variable to attribute, fixes github issue #4

1.10 2018-09-05T16:12:57Z
    - change internals to use helper class object
    - fixes problems with synopsis example reported by tyldis (Vidar Tyldum) via irc

1.01 2018-06-26T15:33:43Z
    - use non-leaking version of private callback method
    - restore typo fix in synopsis (Dan book)

1.0 2018-06-25T14:41:53Z

    - BREAKING CHANGE: rename "stop_queue" event to "queue_empty"
    - fix breakage with Mojolicious v7.85 caused by having a method called _process() that conflicts with a new method in Mojo::UserAgent

0.04 2018-01-01T17:23:07Z

    - refactor internals, replace timer with events 
    - remove attribute "active"
    - updated required perl version
    - mention promises in synopsis

0.03 2018-01-01T14:07:28Z

    - added missing Class::Method::Modifiers dependency
    - fix bug where more than one timer was created

0.02 2017-12-31T14:19:21Z

    - added missing Role::Tiny dependency

0.01 2017-12-30T21:29:07Z

    - original version



lib/Mojo/UserAgent/Role/Queued/Queue.pm
package Mojo::UserAgent::Role::Queued::Queue;

use strict;
use warnings;

use Mojo::Base 'Mojo::EventEmitter';

has jobs => sub { [] };
has active => 0;
has max_active => 4;
has original_method => sub { return sub { die "No Method set!"; } };


sub process {
   my ($self) = @_;
  # we have jobs and can run them:
  while ($self->active < $self->max_active
    and my $job = shift @{$self->jobs})
  {
    my ($tx, $cb) = ($job->{tx}, $job->{cb});
    $self->active($self->active + 1);
    weaken $self;
    $tx->on(finish => sub { $self->active($self->active - 1); $self->process() });
    $self->original_method->( $self, $tx, $cb );
  }
  if (scalar @{$self->jobs} == 0 && $self->active == 0) {
    $self->emit('queue_empty');
  }
}

sub enqueue {
    my ($self, $job) = @_;
    push @{$self->jobs}, $job;
    $self->process();
}

1;


lib/Mojo/UserAgent/Role/Queued.pm
package Mojo::UserAgent::Role::Queued;
use Mojo::Base '-role';

our $VERSION = "1.12";
use Mojo::UserAgent::Role::Queued::Queue;

has max_active => sub { shift->max_connections };

has queue => sub {
  Mojo::UserAgent::Role::Queued::Queue->new(max_active => shift->max_active);
};

around start => sub {
  my ($orig, $self, $tx, $cb) = @_;
  $self->queue->callback(sub { $self->$orig(@_) })
    unless ($self->queue->callback);
  if ($cb) {
    weaken $self;
    $tx->on(finish => sub { $self->queue->tx_finish(); });
    $self->queue->on(queue_empty => sub { $self->emit('queue_empty') });
    $self->queue->enqueue([$tx, $cb]);
  }
  else {
    return $orig->($self, $tx);    # Blocking calls skip the queue
  }
};


1;
__END__

=encoding utf-8

=head1 NAME

Mojo::UserAgent::Role::Queued - A role to process non-blocking requests in a rate-limiting queue.

=head1 SYNOPSIS

    use Mojo::UserAgent;

    my $ua = Mojo::UserAgent->new->with_roles('+Queued');
    $ua->max_redirects(3);
    $ua->max_active(5); # process up to 5 requests at a time
    for my $url (@big_list_of_urls) {
    $ua->get($url, sub {
            my ($ua, $tx) = @_;
            if ($tx->success) {
                say "Page at $url is titled: ",
                  $tx->res->dom->at('title')->text;
            }
           });
   };
   # works with promises, too:
  my @p = map {
    $ua->get_p($_)->then(sub { pop->res->dom->at('title')->text })
      ->catch(sub { say "Error: ", @_ })
  } @big_list_of_urls;
   Mojo::Promise->all(@p)->wait;
 

=head1 DESCRIPTION

Mojo::UserAgent::Role::Queued manages all non-blocking requests made through L<Mojo::UserAgent> in a queue to limit the number of simultaneous requests.

L<Mojo::UserAgent> can make multiple concurrent non-blocking HTTP requests using Mojo's event loop, but because there is only a single process handling all of them, you must take care to limit the number of simultaneous requests you make.

Some discussion of this issue is available here
L<http://blogs.perl.org/users/stas/2013/01/web-scraping-with-modern-perl-part-1.html>
and in Joel Berger's answer here:
L<http://stackoverflow.com/questions/15152633/perl-mojo-and-json-for-simultaneous-requests>.

L<Mojo::UserAgent::Role::Queued> tries to generalize the practice of managing a large number of requests using a queue, by embedding the queue inside L<Mojo::UserAgent> itself.

=head1 EVENTS

L<Mojo::UserAgent::Role::Queued> adds the following event to those emitted by L<Mojo::UserAgent>:

=head2 queue_empty

  $ua->on(queue_empty => sub { my ($ua) = @_; .... })

Emitted when the queue has been emptied of all pending jobs. In previous releases, this event was called C<stop_queue> (B<this is a breaking change>).

=head1 ATTRIBUTES

L<Mojo::UserAgent::Role::Queued> has the following attributes:

=head2 max_active

    $ua->max_active(5);  # execute no more than 5 transactions at a time.
    print "Execute no more than ", $ua->max_active, " concurrent transactions"

Parameter controlling the maximum number of transactions that can be active at the same time.

=head2 

=head1 LICENSE

Copyright (C) Dotan Dimet.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Dotan Dimet E<lt>dotan@corky.netE<gt>

=cut

